Add Caffeine implementation for SessionStore

pendula95 opened the pull request:
Fixes #2692

tsegismont commented on vertx-web-session-stores/vertx-web-sstore-caffeine/src/main/java/io/vertx/ext/web/sstore/caffeine/CaffeineSessionStore.java:
```diff
@@ -0,0 +1,50 @@
+/**
+ * A session store based on the Caffeine which is only available on a single node.
+ * <p>
+ * Can be used when sticky sessions are being used.
```
Perhaps could you add a note saying that sticky sessions is a third-party proxy feature, and that the session in this case won't survive a node crash.

tsegismont commented on vertx-web-session-stores/vertx-web-sstore-caffeine/src/main/java/io/vertx/ext/web/sstore/caffeine/impl/CaffeineSessionStoreImpl.java:
```diff
@@ -0,0 +1,140 @@
+    this.random = VertxContextPRNG.current(vertx);
+    this.vertx = (VertxInternal) vertx;
+    this.sessionMapName = options.getString("mapName", DEFAULT_SESSION_MAP_NAME);
+    final ContextInternal ctx = ((VertxInternal) vertx).getOrCreateContext();
+    localCaffeineCache = (AsyncCache<String, Session>) ctx.contextData()
+      .computeIfAbsent(sessionMapName, k -> Caffeine.newBuilder()
+        .executor(cmd -> ctx.runOnContext(v -> cmd.run()))
+        .expireAfter(Expiry.creating((String key, Session session) ->
+          Duration.ofMillis(session.timeout())))
+        .buildAsync());
+    return this;
```
I don't understand why we can't create an async cache directly, can you explain?

pendula95 replied:
I wanted to store the cache somewhere where it can be accessed from multiple parts of code and reused. This is to achieve the same thing as `vertx.sharedData().getLocalMap` that i used for LocalSessionStoreImpl

tsegismont replied:
Shared data can be accessed from any thread provided you have a Vert.x instamce, context data is only available if you have a reference to the context instance. So I don't believe storing the cache in context data will work, for example if several instances of the verticle are deployed.

Regarding access to the session cache or map from external map, have you got an example use case?

pendula95 replied:
So if I understand correctly if I have 2 vertices deployed ((VertxInternal) vertx).getOrCreateContext(); will not result in same context and hence have different context data? In this case current implementation would have different instances of caffeine cache?  

Use case where this cache needs to be shared is for example deploying multiple instances of same Verticle. If cache is not shared between them then session is rendered useless. This is resolved in https://github.com/vert-x3/vertx-web/blob/10fcd1f668dc1aceb0f5d9fb0e4ab4b030b08fb9/vertx-web/src/main/java/io/vertx/ext/web/sstore/impl/LocalSessionStoreImpl.java#L77

Issues is that Caffeine caches can not be put in shareData

tsegismont replied:
Yes, contexts (thus related data) would be different in different instances of a verticle.

Right, that use case you describe is actually what's needed for the `LocalSessionStore` implementation. But other than that, I'm not sure why we'd need to access the shared local map from the outside.

pendula95 replied:
There is no need to access it outside of the LocalSessionImpl, there is just need to share it across different Vertices.

tsegismont replied:
Ok, then I think you could use `io.vertx.core.impl.VertxImpl#createSharedResource`

tsegismont reviewed (commented):
Thanks for this PR @pendula95 

Just did a quick pass and I have a question about cache creation.

tsegismont reviewed (commented):
Thanks for the updates @pendula95 

I think the implementation could be simplified using a synchronous cache. Indeed, we don't use cache loading, so the cache is sort of a concurrent maps on steroids.

pendula95 commented on vertx-web-session-stores/vertx-web-sstore-caffeine/src/main/java/io/vertx/ext/web/sstore/caffeine/impl/CaffeineSessionStoreImpl.java:
```diff
@@ -0,0 +1,147 @@
+    localCaffeineCache = this.vertx.createSharedResource("__vertx.shared.caffeine.sessions.store", sessionCacheName, closeFuture, cf_ -> {
+      AsyncCache<String, Session> localCaffeineCache = Caffeine.newBuilder()
+        .executor(cmd -> ctx.runOnContext(v -> cmd.run()))
+        .expireAfter(Expiry.creating((String key, Session session) ->
```
You are right, I missed the implementation of cleaning code (inside handle method) that checks last accessed time to check if it should be deleted.

pendula95 commented:
> Thanks for the updates @pendula95
> 
> I think the implementation could be simplified using a synchronous cache. Indeed, we don't use cache loading, so the cache is sort of a concurrent maps on steroids.

Thank you for the input @tsegismont . I was also in dilemma myself if async cache should be used as we do not have any loading for it (it is used directly as simple cache). I have switched to sync cache. I also removed execute blocking code on the few of the methods where I have previously used it. Parts of the code should not block but there are cases where ConcurentHashMap might block but those are marginal. I checked other parts of vert.x code where similar structure is used and did not find execute blocking wrappers.

Tests are now failing but on the part that unrelated to the code changes of this PR, or this is my bad assumption?

tsegismont commented:
> Parts of the code should not block but there are cases where ConcurentHashMap might block but those are marginal. I checked other parts of vert.x code where similar structure is used and did not find execute blocking wrappers.

Exactly

> Tests are now failing but on the part that is unrelated to the code changes of this PR, or this is my bad assumption?

All checks are green, perhaps are you talking about a previous execution?

tsegismont commented on vertx-web-session-stores/vertx-web-sstore-caffeine/src/main/java/io/vertx/ext/web/sstore/caffeine/impl/CaffeineSessionStoreImpl.java:
```diff
@@ -0,0 +1,131 @@
+    CloseFuture closeFuture = new CloseFuture();
+    localCaffeineCache = this.vertx.createSharedResource("__vertx.shared.caffeine.sessions.store", sessionCacheName, closeFuture, cf_ -> {
+      Cache<String, Session> localCaffeineCache = Caffeine.newBuilder()
+        .executor(cmd -> ctx.runOnContext(v -> cmd.run()))
```
```suggestion
```

I believe this is only needed for loading tasks or notifications, which we don't use.

pendula95 replied:
I have considered removing this also (same story sync vs async) but then I read this:
https://github.com/ben-manes/caffeine/blob/701df97f7597ffec5ce772014ed43b963e56a453/caffeine/src/main/java/com/github/benmanes/caffeine/cache/Caffeine.java#L329-L347


There are maintenance task that will be run via this context, probably eviction so I left this so that it is executed on already existing event loop instead of using default ForkJoinPool.

tsegismont replied:
Good point, thanks

tsegismont reviewed (approved):
This looks good to me, with one suggestion

Thanks @pendula95 !

tsegismont commented:
@vietj this is ready to merge. I'll wait a few days before merging in case you want to take a look as well.

vietj commented:
looks great thanks @pendula95
